druid/src/menu/mod.rs

```diff
@@ -0,0 +1,762 @@
+            build: Box::new(build),
+            popup: false,
+            old_data: None,
+            menu: None,
+        }
+    }
+
+    /// Create a new [`MenuManager`] for a context menu.
+    pub fn new_for_popup(build: impl FnMut(&T, &Env) -> Menu<T> + 'static) -> MenuManager<T> {
+        MenuManager {
+            build: Box::new(build),
+            popup: true,
+            old_data: None,
+            menu: None,
+        }
+    }
+
+    /// If this platform always expects windows to have a menu by default, returns a menu.
+    /// Otherwise, returns `None`.
+    #[allow(unreachable_code)]
+    pub fn platform_default() -> Option<MenuManager<T>> {
+        #[cfg(target_os = "macos")]
+        return Some(MenuManager::new(|_, _| sys::mac::application::default()));
+
+        #[cfg(any(target_os = "windows", target_os = "linux"))]
+        return None;
+
+        // we want to explicitly handle all platforms; log if a platform is missing.
+        tracing::warn!("MenuManager::platform_default is not implemented for this platform.");
+        None
+    }
+
+    /// Called when a menu event is received from the system.
+    pub fn event(&mut self, queue: &mut CommandQueue, id: MenuItemId, data: &mut T, env: &Env) {
+        if let Some(m) = &mut self.menu {
+            let mut ctx = MenuEventCtx { queue };
+            m.activate(&mut ctx, id, data, env);
+        }
+    }
+
+    /// Build an initial menu from the application data.
+    pub fn initialize(&mut self, data: &T, env: &Env) -> PlatformMenu {
+        self.menu = Some((self.build)(data, env));
+        self.old_data = Some(data.clone());
+        self.refresh(data, env)
+    }
+
+    /// Update the menu based on a change to the data.
+    ///
+    /// Returns a new `PlatformMenu` if the menu has changed; returns `None` if it hasn't.
+    pub fn update(&mut self, data: &T, env: &Env) -> Option<PlatformMenu> {
+        if self.old_data.is_none() || self.menu.is_none() {
+            tracing::error!("tried to update uninitialized menus");
+            return None;
+        }
+
+        let old_data = self.old_data.as_ref().unwrap();
+        let menu = self.menu.as_mut().unwrap();
+        let ret = match menu.update(old_data, data, env) {
+            MenuUpdate::NeedsRebuild => {
+                self.menu = Some((self.build)(data, env));
+                Some(self.refresh(data, env))
+            }
+            MenuUpdate::NeedsRefresh => Some(self.refresh(data, env)),
+            MenuUpdate::UpToDate => None,
+        };
+        self.old_data = Some(data.clone());
+        ret
+    }
+
+    pub fn refresh(&mut self, data: &T, env: &Env) -> PlatformMenu {
+        if self.old_data.is_none() || self.menu.is_none() {
+            tracing::error!("tried to refresh uninitialized menus");
+            return PlatformMenu::new();
+        }
+
+        let mut ctx = MenuBuildCtx::new(self.popup);
+        self.menu
+            .as_mut()
+            .unwrap()
```

Comment:
this is purely academic, but lately I've been curious about the possible code-size costs of unwraps, and am wondering if it's worth trying to avoid them in places like this where it's relatively trivial.

This is not a suggestion, I'm just mentioning it, it's something I'd like to try measuring at some point.

Reply:
It's a good question. I actually think I like it better with an "if let", so I got rid of the unwraps anyway. I wouldn't be terribly surprised, though, if rustc is smart enough to remove it in cases like this (https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=9f50ee6b6740c01277b4f680658deb31)